archive/extract.go

```diff
@@ -15,10 +16,35 @@ type PathMode struct {
 	Mode os.FileMode
 }
 
+var (
+	umaskLock      sync.Mutex
+	extractCounter int
+	originalUmask  int
+)
+
+func setUmaskIfNeeded() {
+	umaskLock.Lock()
+	extractCounter++
+	if extractCounter == 1 {
+		originalUmask = setUmask(0)
+	}
+	umaskLock.Unlock()
```

Comment:
I would feel safer with `defer umaskLock.Unlock()` on line 27. Mostly to guard against an unexpected `panic` when setting `umask`.

Reply:
Please see https://github.com/buildpacks/lifecycle/pull/730/commits/901a5d6146ffee65c3eb3a9c460df2446af4f6ec.